### ISSUE TITLE
Regression (?): skipping evaluation until null support is provided

### ISSUE THREAD

**Issue by zarbis** [influxdata/flux#1271]
I've upgraded InfluxDB form 1.7.3 to 1.7.6 and many similar queries started to fail with such error: `failed to evaluate map function: null reference used in row function: skipping evaluation until null support is provided`

Result seems to depend more on current data rather than variations of queries like this one:

```go
from(...)
|> range(...)
|> filter(...)
|> group(columns: ["action"])
|> aggregateWindow(every: 1d, fn: sum, timeSrc: "_start")
|> pivot(rowKey: ["_time"], columnKey: ["action"], valueColumn: "_value")
|> map(fn: (r) => ({  
  _time: r._time,  
  foo: r.["actionA"],
  bar: r.["actionA"] / r.["actionB"],
  ...
}))
```

I see there is [TODO](https://github.com/influxdata/flux/blob/19d7fe7dbb08a9f456fed5b33662f415e19dd733/execute/row_fn.go#L124-L127) around that error text. Should I wait until it gets fixed, or there is a way to band-aid my queries?

**Comment by nathanielc** [comment]
@zarbis We are close to having support for `nulls` in lambdas see https://github.com/influxdata/flux/pull/1278 

I am not aware of a workaround in 1.7.6 for this, @affo Do you have any ideas?

**Comment by affo** [comment]
@zarbis if you are ok with discarding the rows that generate the error, I think a workaround is to fill in null values and filter out matching rows.

The objective is not to pass any record that contains null values that your mapping function uses. That would cause the evaluation to fail.

Supposing you don't have null timestamps, I suppose that the problem lies in `actionA` and `actionB`.
What you could do is:

```
specialV = "---NA---"

from(...)
|> range(...)
|> filter(...)
|> group(columns: ["action"])
|> aggregateWindow(every: 1d, fn: sum, timeSrc: "_start")
|> pivot(rowKey: ["_time"], columnKey: ["action"], valueColumn: "_value")

// here's the new part
|> fill(column: "actionA", value: specialV)
|> fill(column: "actionB", value: specialV)
|> filter(fn: (r) => r.actionA != specialV and r.actionB != specialV)

|> map(fn: (r) => ({  
  _time: r._time,  
  foo: r.actionA,
  bar: r.actionA / r.actionB,
  ...
}))
```

Once #1270 gets merged and a new version of Flux gets released, you'll be able to use conditional logical to test for null values and take different decisions based on that.

**Comment by zarbis** [comment]
@affo How should I work around problem, when I expect `pivot()` to produce `foo|bar|baz` columns to use them later on in `map()`, but some time interval contains only rows with `action=foo|bar` and no `action=baz` events happened.

Doing
```js
|> set(column: foo, value: "dummy")
|> set(column: bar, value: "dummy")
|> set(column: baz, value: "dummy")
|> pivot(rowKey: ["_time"], columnKey: ["action"], valueColumn: "_value")
```
to ensure all columns exist hits strange limitation of set only working with `string` type so I cannot set zeros.

And doing `fill()-after-pivot()` on non-existent column doesn't work either.

**Comment by affo** [comment]
@zarbis are you still using 1.7.6? I think you are using Flux v0.24 or prior. At that version you don't have any conditional logic (`if a then b else c`) nor existence check (`exists r.baz`), so I suppose you cannot have a clean work around on that, I am sorry.

When you have a missing column you can only get errors back when you try to access it, and `set` shouldn't work because it doesn't set conditionally and it would overwrite all your `_value`s!

I suppose the only workaround possible is to ensure that you have at least one record that has `action` set to `foo`, one to `bar`, and one to `buz`.

You can do this by:

```
data = () => {...} // suppose this is the function that provides you data

// your data
d = data()
 // add a column to understand if I should use this record
 |> set(column: "use", value: "yes")

// this result will have the same schema as your data and contain one row for each table
// that contains the action "foo" and an invalid value (use == "no").
foo = data()
 |> limit(n: 1)
 |> set(column: "action", value: "foo")
 |> set(column: "use", value: "no")

// same for "bar"
bar = data()
 |> limit(n: 1)
 |> set(column: "action", value: "bar")
 |> set(column: "_value", value: "don't use me")

// same for "baz"
baz = data()
 |> limit(n: 1)
 |> set(column: "action", value: "baz")
 |> set(column: "_value", value: "don't use me")

// ok now we can unify our tables.
// At this point we are sure that every table has every action.
union(tables: [data(), foo, bar, baz])
 |> pivot(...)
 // we can filter out what can be used.
 |> filter(fn: (r) => r.use == "yes")
```

**Comment by zarbis** [comment]
@affo this one works for me. Functions you've mentioned (conditionals, `exists`) are coming in next releases?

**Comment by affo** [comment]
@zarbis glad the workaround works!

Yes, that functions are available now in InfluxDB 2.x. They have also been ported to 1.x (https://github.com/influxdata/influxdb/issues/14342) but they are not available (still) via standard installation. Once new 1.x updates will be published (1.7.8 I suppose) they will be available (along with breaking changes in Flux language 😄
